application/classes/Ushahidi/Formatter/Post/CSV.php

```diff
@@ -34,25 +35,52 @@ class Ushahidi_Formatter_Post_CSV extends Ushahidi_Formatter_API
 	protected $search;
 	protected $fs;
 	protected $tmpfname;
-	protected $add_header;
+	protected $add_header = true;
+	protected $heading;
 
 	// Formatter
-	public function __invoke($records)
+	public function __invoke($records, $attributes = [])
 	{
-		return $this->generateCSVRecords($records);
+		if ($this->heading) {
```

Comment:
So headers are set for the entire CSV and then we call `invoke` for each line? Am I following this all correctly?

Reply:
> So headers are set for the entire CSV 
YES
> and then we call invoke for each line?
NOPE. This is called once `
		$file = service("formatter.entity.post.$format")->__invoke($posts, $keyAttributes);`

Reply:
Thats weird then. Why don't we just pass headers as a param with all the posts?

Reply:
oh. I should add a comment if I didn't . The headers var is good for output (has the correct columns) but it's useless for the processing tasks we need to do in the formatter (they don't have type and other attrs that $keyAttributes has)